### PR TITLE
Support 'href' dependencies 

### LINKED ISSUE #794: validate requires file but does not check href
While developing `googlefontR`, I discovered that [lines](https://github.com/rstudio/rmarkdown/blob/master/R/html_dependencies.R#L217-L220) only check for `file` and not `href` when validating dependencies.  It seems either `file` or `href` should make a valid dependency.  I would like to not have to download every font available from Google.  Is it possible to validate with just `href` as `src`?

### LINKED ISSUE #2021: Error: path for html_dependency not provided.
---
Hi, 

I'm trying to render a Rmarkdown file in linux and eventually I want it to generate a dynamic .html file for people to checkout. The command I'm using is

`R -e "rmarkdown::render('./test_rshiny2.Rmd',output_file='./results/Diagnos_Plot.html')".`

However I kept getting this error:

`Error: path for html_dependency not provided.`

Below is my the session info:
```
> xfun::session_info('rmarkdown')
R version 3.6.3 (2020-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.3 LTS

Locale:
  LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
  LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
  LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
  LC_PAPER=en_US.UTF-8       LC_NAME=C                 
  LC_ADDRESS=C               LC_TELEPHONE=C            
  LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

Package version:
  base64enc_0.1.3 digest_0.6.25   evaluate_0.14   glue_1.4.2     
  graphics_3.6.3  grDevices_3.6.3 highr_0.8       htmltools_0.5.0
  jsonlite_1.7.1  knitr_1.30      magrittr_1.5    markdown_1.1   
  methods_3.6.3   mime_0.9        rlang_0.4.8     rmarkdown_2.4  
  stats_3.6.3     stringi_1.5.3   stringr_1.4.0   tinytex_0.26   
  tools_3.6.3     utils_3.6.3     xfun_0.18       yaml_2.2.1     

Pandoc version: 2.7.3
```

Here is the stack over flow link of my question:
https://stackoverflow.com/questions/65877590/render-rmarkdown-file-in-command-line-got-error-path-for-html-dependency-not-pr

Thanks very much for your help!!

By filing an issue to this repo, I promise that

- [X] I have fully read the issue guide at https://yihui.org/issue/.
- [X] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('rmarkdown')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('rstudio/rmarkdown')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

### LINKED ISSUE #2147: create a widget without an R package
Follow [https://bookdown.org/yihui/rmarkdown/htmlwidgets-advanced.html#create-a-widget-without-an-r-package](https://bookdown.org/yihui/rmarkdown/htmlwidgets-advanced.html#create-a-widget-without-an-r-package)

the test.Rmd as follows：
````R
---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,error = T)
```

## R Markdown

```{r}
library(htmlwidgets)
blink = function(text, interval = 1) {
  htmlwidgets::createWidget(
    'blink', 
    list(text = text, interval = interval),
    dependencies = htmltools::htmlDependency(
      'blink', '0.1', src = c(href = ''), head = '
<script>
HTMLWidgets.widget({
  name: "blink",
  type: "output",
  factory: function(el, width, height) {
    return {
      renderValue: function(x) {
        setInterval(function() {
          el.innerText = el.innerText == "" ? x.text : "";
        }, x.interval * 1000);
      },
      resize: function(width, height) {}
    };
  }
});
</script>'
    ),
  )
}

blink('Hello htmlwidgets!', .5)
````
when I use `render("test.Rmd")`,An error occurred：`Error: path for html_dependency not provided`
<details>
  <summary>xfun::session_info()</summary>
  
```R
R version 4.1.0 (2021-05-18)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.2 LTS, RStudio 1.4.1106

Locale:
  LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
  LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
  LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
  LC_PAPER=en_US.UTF-8       LC_NAME=C                 
  LC_ADDRESS=C               LC_TELEPHONE=C            
  LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

Package version:
  base64enc_0.1.3   bslib_0.2.5.1     compiler_4.1.0   
  digest_0.6.27     evaluate_0.14     fs_1.5.0         
  glue_1.4.2        graphics_4.1.0    grDevices_4.1.0  
  highr_0.9         htmltools_0.5.1.1 htmlwidgets_1.5.3
  jquerylib_0.1.4   jsonlite_1.7.2    knitr_1.33       
  magrittr_2.0.1    markdown_1.1      methods_4.1.0    
  mime_0.10         R6_2.5.0          rappdirs_0.3.3   
  rlang_0.4.11      rmarkdown_2.8     sass_0.4.0       
  stats_4.1.0       stringi_1.6.2     stringr_1.4.0    
  tinytex_0.31      tools_4.1.0       utils_4.1.0      
  xfun_0.23         yaml_2.2.1 
```
</details>

## Checklist

When filing a _bug report_, please check the boxes below to confirm that you have provided us with the information we need. Have you:

- [x] [formatted your issue](https://yihui.org/issue/#please-format-your-issue-correctly) so it is easier for us to read?

- [x] included a minimal, self-contained, and reproducible example?

- [x] pasted the output from `xfun::session_info('rmarkdown')` in your issue?

- [x] upgraded all your packages to their latest versions (including your versions of R, the RStudio IDE, and relevant R packages)?

- [ ] installed and tested your bug with the development version of the rmarkdown package using `remotes::install_github("rstudio/rmarkdown")`?

### PR DESCRIPTION
**Pull request by cderv**
This PR is a duplicate of #1948 where I can't push. It also superseed #1805. 

This will fix #2021, fix #2147, fix #794 and fix part of #1842

This sum up the work of @matthewstrasiotto and @crazycapivara with addition on my side based on review element. 

Tests were also added to increase the coverage and be sure we did not miss something. 

With this PR, it is now possible to use HTML dependency that have only a `href` in their `src`

<details>
<summary>Like example in #2147</summary>

````markdown
---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,error = T)
```

## R Markdown

```{r}
library(htmlwidgets)
blink = function(text, interval = 1) {
  htmlwidgets::createWidget(
    'blink', 
    list(text = text, interval = interval),
    dependencies = htmltools::htmlDependency(
      'blink', '0.1', src = c(href = ''), head = '
<script>
HTMLWidgets.widget({
  name: "blink",
  type: "output",
  factory: function(el, width, height) {
    return {
      renderValue: function(x) {
        setInterval(function() {
          el.innerText = el.innerText == "" ? x.text : "";
        }, x.interval * 1000);
      },
      resize: function(width, height) {}
    };
  }
});
</script>'
    ),
  )
}

blink('Hello htmlwidgets!', .5)
```
````

</details>

<details>
<summary>Or like in #1842</summary>

````markdown
---
output: html_document
---

```{r echo=FALSE}
library(htmltools)
```

```{r}
tagList(htmlDependency(
  name = "google-font-exo-2",
  version = "1.0.0",
  src = c(href = "https://fonts.googleapis.com/"),
  stylesheet = "css2?family=Exo+2:ital,wght@0,100;0,200;0,300;0,400;0,500;0,600;0,700;0,800;0,900;1,100;1,200;1,300;1,400;1,500;1,600;1,700;1,800;1,900&display=swap"
))
tags$style("body { font-family: 'Exo 2'; }")
```

Some content

````
</details>

### CODE REVIEW

**Review by yihui — COMMENTED** [review]
That's fantastic! Thanks!

File: R/html_dependencies.R [hunk 1]
```diff
@@ -328,14 +343,18 @@ validate_html_dependency <- function(list) {
...
-  file <- list$src$file
-  if (!is.null(list$package))
-    file <- system.file(file, package = list$package)
-  if (!file.exists(file)) {
-    utils::str(list)
-    stop("path for html_dependency not found: ", file, call. = FALSE)
+
+  # check src path or href are given
+  if (is.null(list$src)) {
+    stop("neither path nor href are provided for html_dependency", call. = FALSE)
```
Comment (yihui) [inline R/html_dependencies.R:349]: Currently we have a lot of `call. = FALSE` in this package. Perhaps we should copy these two helper functions: https://github.com/yihui/knitr/blob/89a1f6cc19e060c0f45b80fe1c4905dfab441193/R/utils.R#L870-L871

Comment (cderv) [reply R/html_dependencies.R:349]: Yes I was looking for them in **xfun** at first and saw that they are internal to **knitr**. 
I can surely copy them and replace every call !

Comment (yihui) [reply R/html_dependencies.R:349]: They are one-liners and probably not worth exporting in **xfun**.

**Pull request merged**